File: kotlinapp/app/src/main/java/com/emusic/app/ui/home/HomeViewModel.kt
package com.emusic.app.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.emusic.app.data.api.GenreStat
import com.emusic.app.data.api.Track
import com.emusic.app.data.repository.MusicRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class HomeUiState(
    val mostPlayed: List<Track> = emptyList(),
    val recommendations: List<Track> = emptyList(),
    val topGenres: List<GenreStat> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingReco: Boolean = false,
    val genreResults: List<Track> = emptyList(),
    val loadingGenre: String? = null,
    val error: String? = null
)

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val repository: MusicRepository
) : ViewModel() {

    private val _state = MutableStateFlow(HomeUiState())
    val state: StateFlow<HomeUiState> = _state.asStateFlow()

    init {
        loadAll()
    }

    fun loadAll() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true)
            val mostPlayed = repository.getMostPlayed(20)
            val genres = repository.getTopGenres()
            _state.value = _state.value.copy(
                mostPlayed = mostPlayed,
                topGenres = genres,
                isLoading = false
            )
        }
        loadRecommendations()
    }

    private fun loadRecommendations() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoadingReco = true)
            val tracks = repository.getRecommendations()
            _state.value = _state.value.copy(recommendations = tracks, isLoadingReco = false)
        }
    }

    fun searchGenre(genreName: String, query: String, onReady: (List<Track>) -> Unit) {
        viewModelScope.launch {
            _state.value = _state.value.copy(loadingGenre = genreName)
            val results = repository.search(query)
            _state.value = _state.value.copy(genreResults = results, loadingGenre = null)
            if (results.isNotEmpty()) onReady(results)
        }
    }

    /** "No me interesa": excluye el track de futuras recomendaciones y lo quita ya. */
    fun excludeRecommendation(track: Track) {
        _state.value = _state.value.copy(
            recommendations = _state.value.recommendations.filterNot { it.videoId == track.videoId }
        )
        viewModelScope.launch {
            repository.excludeRecommendation(track)
        }
    }

    fun refresh() {
        loadAll()
    }
}

File: kotlinapp/iosApp/iosApp/HomeViewModel.swift
import Foundation
import shared

/// Puerto de `ui/home/HomeViewModel.kt`. Carga las tres secciones del Inicio en paralelo
/// contra los mismos endpoints que Android, a traves del `MusicApiClient` compartido.
@MainActor
final class HomeViewModel: ObservableObject {

    @Published private(set) var mostPlayed: [Track] = []
    @Published private(set) var recommendations: [Track] = []
    @Published private(set) var topGenres: [GenreStat] = []
    @Published private(set) var isLoading = false
    @Published private(set) var isLoadingReco = false

    /// Fallo de red al cargar Inicio. Se pinta bajo la cabecera.
    @Published private(set) var errorMessage: String?

    /// Nombre del genero cuya busqueda esta en curso, para poner esa tarjeta en "Cargando...".
    @Published private(set) var loadingGenre: String?

    private let api = SharedClients.shared.api
    private let network = SharedClients.shared.network

    /// Cuantas recomendaciones mostrar. Con el tope de 2 por artista que aplica el backend,
    /// 12 garantiza al menos 6 artistas distintos.
    private let recommendationLimit: Int32 = 12

    var isEmpty: Bool {
        !isLoading && !isLoadingReco && mostPlayed.isEmpty && recommendations.isEmpty
    }

    func loadAll() async {
        async let historyAndGenres: Void = loadHistoryAndGenres()
        async let reco: Void = loadRecommendations()
        _ = await (historyAndGenres, reco)
    }

    private func loadHistoryAndGenres() async {
        isLoading = true
        defer { isLoading = false }

        // "Mas escuchadas" no es un endpoint propio: es el historial ordenado por
        // playCount, igual que hace MusicRepository.getMostPlayed en Android.
        //
        // Los errores NO se tragan en silencio: una lista vacia por un fallo de red se
        // ve igual que una lista vacia de verdad, y eso ya costo una tarde de depuracion.
        do {
            let history = try await api.getHistory()
            mostPlayed = history
                .sorted { $0.playCount > $1.playCount }
                .prefix(20)
                .map { $0.toTrack() }
            errorMessage = nil
        } catch {
            errorMessage = "No se pudo conectar con el servidor: \(error.localizedDescription)"
        }

        if let genres = try? await api.getTopGenres() {
            topGenres = genres
        }
    }

    private func loadRecommendations() async {
        isLoadingReco = true
        defer { isLoadingReco = false }

        if let response = try? await api.getRecommendations(limit: recommendationLimit) {
            recommendations = response.items
        }
    }

    /// Tocar un genero dispara la MISMA busqueda que usa el buscador y arranca en modo
    /// radio con el primer resultado — no encola la lista cruda (ver RadioEngine).
    /// Devuelve true si arrancó a sonar, para que la pantalla abra el reproductor
    /// (el `onTrackClick` de `HomeScreen.kt`, que navega a Screen.Player).
    @discardableResult
    func playGenre(_ genre: EMusicGenre, player: PlayerEngine) async -> Bool {
        loadingGenre = genre.name
        defer { loadingGenre = nil }

        guard let results = try? await network.search(query: genre.query),
              let first = results.first else { return false }

        // Nada de prefetch aca. `HomeViewModel.searchGenre` en Android tampoco lo hace, y
        // por una razon: el backend NO deduplica peticiones en vuelo, asi que precargar el
        // mismo videoId que el reproductor esta a punto de pedir lanza yt-dlp DOS VECES en
        // paralelo en la Pi y el arranque tarda el doble. El prefetch tiene sentido cuando
        // hay un hueco antes del play (el buscador, o los siguientes de la cola).
        player.play(track: first, radioSeed: true)
        return true
    }

    /// "No me interesa": lo saca de la lista al instante y lo excluye en el backend.
    func exclude(_ track: Track) {
        recommendations.removeAll { $0.videoId == track.videoId }
        Task {
            try? await api.excludeRecommendation(
                request: ExcludeRequest(
                    videoId: track.videoId,
                    artist: track.displayArtist,
                    title: track.title
                )
            )
        }
    }
}
